File: GoDataProcessing/result_writer.go
package main

import (
	"os"
	"sync"
)

type ResultWriter struct {
	file *os.File
	mu   sync.Mutex
}

func NewResultWriter(filename string) (*ResultWriter, error) {
	f, err := os.Create(filename)
	if err != nil {
		return nil, err
	}
	return &ResultWriter{file: f}, nil
}

func (w *ResultWriter) Write(result string) {
	w.mu.Lock()
	defer w.mu.Unlock()
	w.file.WriteString(result + "\n")
}

func (w *ResultWriter) Close() {
	w.file.Close()
}

File: DataProcessingSystemJava/src/main/java/com/example/dataprocessing/service/ResultWriter.java
// ResultWriter.java
package com.example.dataprocessing.service;

import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class ResultWriter {
    private final PrintWriter writer;
    private final Lock lock = new ReentrantLock();

    public ResultWriter(String filename) throws IOException {
        this.writer = new PrintWriter(new FileWriter(filename, true));
    }

    public void writeResult(String result) {
        lock.lock();
        try {
            writer.println(result);
            writer.flush();
        } finally {
            lock.unlock();
        }
    }

    public void close() {
        writer.close();
    }
}